### PR TITLE
[Communication] - #2kb2z4b - [Library][Span and Trace] SMS Jatis Adjusment

### PR DESCRIPTION
**Pull request by fairyhunter13**
# Changelog
- Add editor configs and env file to .gitignore 🙈
- Add unknown error for status code above or equal 400 ✨
- Update go modules 📦
- Using functional options ✨
- Modify using heimdall.Doer for the HTTP client ✨
- Add default feature for payload ✨
- Implement error interface for response ✨
- Change channel type from uint to uint64 ✨
- Completing the integration testing ✅✨
- Updating go modules 📦
- Add .env.example for integration testing ✨

### CODE REVIEW

**Comment by pianzm** [comment]
@hafiz-ludyanto this seems a huge change in terms of the code base. Possible to add unit tests as well?

**Comment by fairyhunter13** [comment]
@pianzm There is an integration test already added. The integration test is more reliable than unit tests since we can't detect any possible errors from providers if only relying on unit tests. The demo of integration tests is already provided in this thread: https://flipid.slack.com/archives/C03M88WCBGT/p1658974304392559.

**Comment by pianzm** [comment]
> @pianzm There is an integration test already added. The integration test is more reliable than unit tests since we can't detect any possible errors from providers if only relying on unit tests. The demo of integration tests is already provided in this thread: https://flipid.slack.com/archives/C03M88WCBGT/p1658974304392559.

Both are important, so I suggest making some unit tests as this repository is public and will have the possibilities to be used by someone else.

**Comment by fairyhunter13** [comment]
> > @pianzm There is an integration test already added. The integration test is more reliable than unit tests since we can't detect any possible errors from providers if only relying on unit tests. The demo of integration tests is already provided in this thread: https://flipid.slack.com/archives/C03M88WCBGT/p1658974304392559.
> 
> Both are important, so I suggest making some unit tests as this repository is public and will have the possibilities to be used by someone else.

@pianzm The problem is that we don't know the exact behaviour of the `SMS Jatis` provider. Since we don't know the actual behaviour of the `SMS Jatis` provider and don't maintain their source code, it's better to stick with the integration testing. This analogy is like making a SQL driver for a database. It's best to be tested with the database directly rather than mimicking the behaviour of the database itself.

**Comment by asqolani-flip** [comment]
> 

just want to ask mas, if we implement unit testing, what do you think about the effort? if the big effort to implement that in this library, AFAIK since currently all lib just implement the integration testing, i think we can put the unit testing to the initiative task for all lib? wdyt mas @pianzm ?

**Comment by fairyhunter13** [comment]
> > 
> 
> just want to ask mas @hafiz-ludyanto , if we implement unit testing, what do you think about the effort? if the big effort to implement that in this library, AFAIK since currently all lib just implement the integration testing, i think we can put the unit testing to the initiative task for all lib? wdyt mas @pianzm ?

@asqolani-flip I think it takes a significant effort to maintain unit testing. Since the provider's API behaviour can change anytime, unit testing doesn't ensure our library works well when connecting to the provider's API. Unit testing also can't detect when the provider's API behaviour changed. So, even though our unit tests passed, it doesn't ensure that our library works when connecting to the provider's API. It also means that our library would get an error if it doesn't conform to the provider's API when it is changed. My suggestion is that if we want to make the unit tests for this library, I think we should ensure that the provider's API doesn't change very often, and we can plan to create a new task to make unit tests for the providers that don't have many changes. This is also becoming one of the reasons why we can't develop the `ValueFirst` Whatsapp library without connecting to their platform first.

File: payload.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+// ChannelType specifies the channel type used in this package.
+type ChannelType uint64
+
+// List of channel used in Jatis.
+const (
+	ChannelRegular ChannelType = iota
+	ChannelAlert
+	ChannelOTP
+)
```
Comment (asqolani-flip) [inline payload.go:18]: mas ini requirement dari jatis bukan ya?

Comment (fairyhunter13) [reply payload.go:18]: @asqolani-flip ya mas, channel yang diset default untuk `Flip` sepertinya menggunakan `ChannelOTP`

**Comment by asqolani-flip** [comment]
> @asqolani-flip I think it takes a significant effort to maintain unit testing. Since the provider's API behaviour can change anytime, unit testing doesn't ensure our library works well when connecting to the provider's API. Unit testing also can't detect when the provider's API behaviour changed. So, even though our unit tests passed, it doesn't ensure that our library works when connecting to the provider's API. It also means that our library would get an error if it doesn't conform to the provider's API when it is changed. My suggestion is that if we want to make the unit tests for this library, I think we should ensure that the provider's API doesn't change very often, and we can plan to create a new task to make unit tests for the providers that don't have many changes. This is also becoming one of the reasons why we can't develop the ValueFirst Whatsapp library without connecting to their platform first.

i see got it, waiting from mas @pianzm, could we conclude this for all library mas?

File: payload.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+// Default returns the default request message for Jatis.
+func (r *RequestMessage) Default(o *Option) *RequestMessage {
+	if r.BatchName == "" {
+		r.BatchName = DefaultBatchOTP
+	}
+
+	if r.Channel == nil || *r.Channel > ChannelOTP {
+		if o == nil || o.Channel == nil {
+			r.Channel = getChannelOTP()
```
Comment (fairyhunter13) [inline payload.go:76]: @asqolani-flip `r.Channel` akan diset secara `default` jika isinya `nil`

**Comment by fairyhunter13** [comment]
> > @asqolani-flip I think it takes a significant effort to maintain unit testing. Since the provider's API behaviour can change anytime, unit testing doesn't ensure our library works well when connecting to the provider's API. Unit testing also can't detect when the provider's API behaviour changed. So, even though our unit tests passed, it doesn't ensure that our library works when connecting to the provider's API. It also means that our library would get an error if it doesn't conform to the provider's API when it is changed. My suggestion is that if we want to make the unit tests for this library, I think we should ensure that the provider's API doesn't change very often, and we can plan to create a new task to make unit tests for the providers that don't have many changes. This is also becoming one of the reasons why we can't develop the ValueFirst Whatsapp library without connecting to their platform first.
> 
> i see got it, waiting from mas @pianzm, could we conclude this for all library mas?

@asqolani-flip Sure, I would be glad to have help from mas @pianzm for the unit testing part. 😄

**Comment by pianzm** [comment]
> 

I have submitted the API docs, please have a look

**Comment by fairyhunter13** [comment]
> > 
> 
> I have submitted the API docs, please have a look

@pianzm Sure mas, I await the good news of your contribution to unit testing.

**Comment by fairyhunter13** [comment]
> > > @asqolani-flip I think it takes a significant effort to maintain unit testing. Since the provider's API behaviour can change anytime, unit testing doesn't ensure our library works well when connecting to the provider's API. Unit testing also can't detect when the provider's API behaviour changed. So, even though our unit tests passed, it doesn't ensure that our library works when connecting to the provider's API. It also means that our library would get an error if it doesn't conform to the provider's API when it is changed. My suggestion is that if we want to make the unit tests for this library, I think we should ensure that the provider's API doesn't change very often, and we can plan to create a new task to make unit tests for the providers that don't have many changes. This is also becoming one of the reasons why we can't develop the ValueFirst Whatsapp library without connecting to their platform first.
> > 
> > 
> > i see got it, waiting from mas @pianzm, could we conclude this for all library mas?
> 
> I'm afraid not @asqolani-flip

@asqolani-flip @pianzm Ok, let's conclude this, if we want to add unit tests, we can set it as a tech debt later.

**Pull request merged**